### PR TITLE
feat: change predeploy contract addresses

### PR DESCRIPTION
**Pull request by kangsorang**
# Description

Change predeploy contract addresses.

Includes 8 transactions behaviors for upgrades in MPT blocks. 

Behaves similarly to ecotone upgrades.

One difference is that upgrade transactions are executed in MPT block - 1 instead of MPT block.

### CODE REVIEW

File: op-node/rollup/derive/attributes.go [hunk 1]
```diff
@@ -8,9 +8,9 @@ import (
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/ethereum/go-ethereum/core/types"
 
+	"github.com/ethereum-optimism/optimism/op-bindings/predeploys"
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes.go:11]: How about adding alias for oppredeploys import? (in all places) It's quite confusing if seeing `predeploys.xxx`

File: op-node/rollup/derive/attributes.go [hunk 2]
```diff
@@ -108,6 +108,17 @@ func (ba *FetchingAttributesBuilder) PreparePayloadAttributes(ctx context.Contex
 		}
 	}
 
+	// [Kroma: START]
+	// Include KromaMPT network upgrade transactions in the block before(parent block) of the KromaMPT target block.
+	if ba.rollupCfg.IsKromaMPTParentBlock(nextL2Time) {
+		kromaMPT, err := KromaMPTNetworkUpgradeTransactions()
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes.go:114]:
```suggestion
		mptUpgradeTxs, err := KromaMPTNetworkUpgradeTransactions()
```

File: op-node/rollup/derive/attributes.go [hunk 3]
```diff
@@ -119,19 +130,24 @@ func (ba *FetchingAttributesBuilder) PreparePayloadAttributes(ctx context.Contex
...
+			for i, otx := range txs {
+				if otx[0] != types.DepositTxType {
+					continue
+				}
+				tx, err := ToKromaDepositBytes(otx)
+				if err != nil {
+					return nil, NewCriticalError(err)
+				}
+				txs[i] = tx
 			}
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes.go:150]:
```suggestion
		for i, otx := range txs {
			if otx[0] != types.DepositTxType {
				continue
			}
			tx, err := ToKromaDepositBytes(otx)
			if err != nil {
				return nil, NewCriticalError(fmt.Errorf("failed to convert deposit tx to KromaDepositTx: %w", err))
			}
			txs[i] = tx
```

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+	deployBaseFeeVaultSource           = UpgradeDepositSource{Intent: "KromaMPT: BaseFee Vault Deployment"}
+	updateBaseFeeVaultProxySource      = UpgradeDepositSource{Intent: "KromaMPT: BaseFee Vault Proxy Update"}
+	deployL1FeeVaultSource             = UpgradeDepositSource{Intent: "KromaMPT: L1 Fee Vault Deployment"}
+	updateL1FeeVaultProxySource        = UpgradeDepositSource{Intent: "KromaMPT: L1 Fee Vault Proxy Update"}
+	deploySequencerFeeVaultSource      = UpgradeDepositSource{Intent: "KromaMPT: Sequencer Fee Vault Deployment"}
+	updateSequencerFeeVaultProxySource = UpgradeDepositSource{Intent: "KromaMPT: Sequencer Fee Vault Proxy Update"}
+
+	deployL1BlockMPTSource       = UpgradeDepositSource{Intent: "KromaMPT: L1 Block Deployment"}
+	updateL1BlockMPTProxySource  = UpgradeDepositSource{Intent: "KromaMPT: L1 Block Proxy Update"}
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:30]:
```suggestion
	deployL1BlockMPTSource       = UpgradeDepositSource{Intent: "KromaMPT: L1 Block Deployment"}
	updateL1BlockMPTProxySource  = UpgradeDepositSource{Intent: "KromaMPT: L1 Block Proxy Update"}

```

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
...248201527f4c31426c6f636b3a206f6e6c7920746865206465706f7369746f72206163636f60448201527f756e742063616e20736574204c3120626c6f636b2076616c7565730000000000606482015260840160405180910390fd5b6000805467ffffffffffffffff98891668010000000000000000027fffffffffffffffffffffffffffffffff00000000000000000000000000000000909116998916999099179890981790975560019490945560029290925560038054919094167fffffffffffffffffffffffffffffffffffffffffffffffff00000000000000009190911617909255600491909155600555600655565b3373deaddeaddeaddeaddeaddeaddeaddeaddead00011461040357633cc50b456000526004601cfd5b60043560801c60035560143560801c600055602435600155604435600755606435600255608435600455565b803567ffffffffffffffff8116811461044757600080fd5b919050565b600080600080600080600080610100898b03121561046957600080fd5b6104728961042f565b975061048060208a0161042f565b9650604089013595506060890135945061049c60808a0161042f565b979a969950949793969560a0850135955060c08501359460e001359350915050565b600060208083528351808285015260005b818110156104eb578581018301518582016040015282016104cf565b818111156104fd576000604083870101525b50601f017fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe01692909201604001939250505056fea164736f6c634300080f000a")
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:32]: Please replace it with actual bytecode

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+)
+
+// init overrides variables defined in the Kroma MPT upgrade.
+func init() {
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:36]: What is it for?

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+func KromaMPTNetworkUpgradeTransactions() ([]hexutil.Bytes, error) {
+	upgradeTxns := make([]hexutil.Bytes, 0, 8)
+
+	deployL1BlockTransaction, err := types.NewTx(&types.DepositTx{
+		SourceHash:          deployL1BlockMPTSource.SourceHash(),
+		From:                L1BlockDeployerAddress,
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:44]:
```suggestion
		From:                KromaMPTDeployerAddress,
```

File: op-node/rollup/derive/kromampt_upgrade_transactions.go (hunk 5 above)
Comment (sm-stack) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:32]: Please change this bytecode to be same as real feeVault contract.

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+	deployBaseFeeVault, err := types.NewTx(&types.DepositTx{
+		SourceHash:          deployBaseFeeVaultSource.SourceHash(),
+		From:                KromaMPTDeployerAddress,
+		To:                  nil,
+		Mint:                big.NewInt(0),
+		Value:               big.NewInt(0),
+		Gas:                 1_000_000,
+		IsSystemTransaction: false,
+		Data:                feeVaultDeploymentBytecode,
+	}).MarshalBinary()
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:68]:
```suggestion
	}).MarshalBinary()
	if err != nil {
		return nil, err
	}
```

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	deployL1FeeVault, err := types.NewTx(&types.DepositTx{
+		SourceHash:          deployL1FeeVaultSource.SourceHash(),
+		From:                KromaMPTDeployerAddress,
+		To:                  nil,
+		Mint:                big.NewInt(0),
+		Value:               big.NewInt(0),
+		Gas:                 1_000_000,
+		IsSystemTransaction: false,
+		Data:                feeVaultDeploymentBytecode,
+	}).MarshalBinary()
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:81]:
```suggestion
	}).MarshalBinary()
		if err != nil {
		return nil, err
	}
```

File: op-node/rollup/derive/span_channel_out.go [hunk 10]
```diff
@@ -102,6 +102,7 @@ func (co *SpanChannelOut) AddSingularBatch(batch *SingularBatch, seqNum uint64)
 	// Append Singular batch to its span batch builder
 	co.spanBatchBuilder.AppendSingularBatch(batch, seqNum)
 	// Convert Span batch to RawSpanBatch
+	//TODO
```
Comment (sm-stack) [inline op-node/rollup/derive/span_channel_out.go:105]: What is this comment for?

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+func KromaMPTNetworkUpgradeTransactions() ([]hexutil.Bytes, error) {
+	upgradeTxns := make([]hexutil.Bytes, 0, 8)
+
+	deployL1BlockTransaction, err := types.NewTx(&types.DepositTx{
+		SourceHash:          deployL1BlockMPTSource.SourceHash(),
+		From:                L1BlockDeployerAddress,
+		To:                  nil,
+		Mint:                big.NewInt(0),
+		Value:               big.NewInt(0),
+		Gas:                 375_000,
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:48]: How is this value calculated?

File: op-node/rollup/derive/kromampt_upgrade_transactions.go (hunk 5 above)
Comment (seolaoh) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:32]: AFAIK, ProtocolVault and L1FeeVault bytecode is slightly different. Maybe we need to separate them

File: op-node/rollup/derive/attributes_queue_test.go [hunk 12]
```diff
@@ -67,10 +73,17 @@ func TestAttributesQueue(t *testing.T) {
 		l2Fetcher := &testutils.MockL2Client{}
 		l2Fetcher.ExpectSystemConfigByL2Hash(safeHead.Hash, parentL1Cfg, nil)
 
+		parentBeaconRoot := l1Info.ParentBeaconRoot()
+		if cfg.IsEcotone(0) {
+			if parentBeaconRoot == nil { // default to zero hash if there is no beacon-block-root available
+				parentBeaconRoot = new(common.Hash)
+			}
+		}
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes_queue_test.go:81]:
```suggestion
		parentBeaconRoot := l1Info.ParentBeaconRoot()
		if cfg.IsEcotone(safeHead.Time+cfg.BlockTime) && parentBeaconRoot == nil { // default to zero hash if there is no beacon-block-root available
			parentBeaconRoot = new(common.Hash)
		}
```

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/predeploys"
+)
+
+var (
+	KromaMPTDeployerAddress = common.HexToAddress("0x2551000000000000000000000000000000000000")
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:15]: I didn't think it was necessary to split it, but it's better to be consistent with ecotone, I'll reflect it.

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+		require.Equal(t, common.HexToHash(test.expectedHash), test.source.SourceHash())
+	}
+}
+
+func TestMPTNetworkTransactions(t *testing.T) {
+	upgradeTxns, err := KromaMPTNetworkUpgradeTransactions()
+	require.NoError(t, err)
+	require.Len(t, upgradeTxns, 8)
+
+	deployL1BlockSender, deployL1Block := toDepositTxn(t, upgradeTxns[0])
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:60]: Please revive below line in `toDepositTxn` function.
```go
	require.False(t, txn.IsSystemTx())
```

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+	require.Equal(t, deployL1BlockSender, L1BlockDeployerAddress)
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:61]:
```suggestion
	require.Equal(t, deployL1BlockSender, KromaMPTDeployerAddress)
```

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+	deployL1FeeVaultSender, deployL1FeeVault := toDepositTxn(t, upgradeTxns[2])
+	require.Equal(t, deployL1FeeVaultSender, KromaMPTDeployerAddress)
+	require.Equal(t, deployL1FeeVaultSource.SourceHash(), deployL1FeeVault.SourceHash())
+	require.Nil(t, deployL1FeeVault.To())
+	require.Equal(t, uint64(1_000_000), deployL1FeeVault.Gas())
+	require.Equal(t, hexutil.Bytes(feeVaultDeploymentBytecode).String(), hexutil.Bytes(deployL1FeeVault.Data()).String())
+
+	deployL1FeeVaultSender, deploySequencerFeeVault := toDepositTxn(t, upgradeTxns[3])
+	require.Equal(t, deployL1FeeVaultSender, KromaMPTDeployerAddress)
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:82]:
```suggestion
	deploySequencerFeeVaultSender, deploySequencerFeeVault := toDepositTxn(t, upgradeTxns[3])
	require.Equal(t, deploySequencerFeeVaultSender, KromaMPTDeployerAddress)
```

File: op-node/rollup/derive/kromampt_upgrade_transactions.go (hunk 11 above)
Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:48]: Referenced the deployL1BlockTransaction gas value from the ecotone upgrade in the OP.

File: op-node/rollup/types.go [hunk 17]
```diff
@@ -371,6 +371,12 @@ func (c *Config) IsKromaMPT(timestamp uint64) bool {
 	return c.KromaMPTTime != nil && timestamp >= *c.KromaMPTTime
 }
 
+// IsKromaMPTParentBlock returns whether the specified block is the parent block subject to the
+// KromaMPT upgrade.
```
Comment (seolaoh) [inline op-node/rollup/types.go:375]:
```suggestion
// KromaMPT upgrade. KromaMPT activation at genesis does not count.
```

File: op-node/rollup/derive/kromampt_upgrade_transactions.go (hunk 11 above)
Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:50]: updated

File: op-node/rollup/derive/fuzz_parsers_test.go [hunk 18]
```diff
@@ -99,6 +99,35 @@ func FuzzL1InfoEcotoneRoundTrip(f *testing.F) {
 	})
 }
 
+// FuzzL1InfoKromaMPTRoundTrip checks that our Kroma MPT encoder round trips properly
+func FuzzL1InfoKromaMPTRoundTrip(f *testing.F) {
+	f.Fuzz(func(t *testing.T, number, time uint64, baseFee, blobBaseFee, hash []byte, seqNumber uint64, baseFeeScalar, blobBaseFeeScalar uint32, validatorRewardScalar []byte) {
```
Comment (seolaoh) [inline op-node/rollup/derive/fuzz_parsers_test.go:104]: don't we need to remove validatorRewardScalar?

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 19]
```diff
@@ -8,48 +8,49 @@ import (
...
+	l1BlockMPTDeploymentBytecode        = common.FromHex(bindings.L1BlockMetaData.Bin)
```
Comment (sm-stack) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:35]: Isn't it better to embed the bytecode itself? If the bytecode changes in the future, I guess it may lead to an unexpected behavior

Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:35]: Yes, I was thinking of readability, but this bytecode should be thought of as a snapshot concept, I'll update.

File: op-node/rollup/derive/fuzz_parsers_test.go (hunk 18 above)
Comment (kangsorang) [reply op-node/rollup/derive/fuzz_parsers_test.go:104]: Yes, that's right, marshaling - unmarshaling behavior, that's why the test called, I'll update it

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 20]
```diff
@@ -32,10 +31,10 @@ var (
...
...0000000000000000000000000000000000000000815260206004820152604360248201527f4c31426c6f636b3a20746865206d61782076616c7565206f662076616c69646160448201527f746f7220726577617264207363616c617220686173206265656e20657863656560648201527f6465640000000000000000000000000000000000000000000000000000000000608482015260a4016103a9565b6000805467ffffffffffffffff998a1668010000000000000000027fffffffffffffffffffffffffffffffff000000000000000000000000000000009091169a8a169a909a179990991790985560019590955560029390935560038054929095167fffffffffffffffffffffffffffffffffffffffffffffffff00000000000000009290921691909117909355600492909255600591909155600655600755565b600060208083528351808285015260005b818110156105385785810183015185820160400152820161051c565b8181111561054a576000604083870101525b50601f017fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe016929092016040019392505050565b803567ffffffffffffffff8116811461059657600080fd5b919050565b60008060008060008060008060006101208a8c0312156105ba57600080fd5b6105c38a61057e565b98506105d160208b0161057e565b975060408a0135965060608a013595506105ed60808b0161057e565b989b979a50959894979660a0860135965060c08601359560e081013595506101000135935091505056fea164736f6c634300080f000a")
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:34]: Why did you update it?

Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:34]: We had this discussion above
https://github.com/kroma-network/kroma/pull/398#discussion_r1881229723
I thought it would be a good idea to put the bytecode in for a snapshot feel.

**Review by sm-stack — APPROVED** [review]

File: op-e2e/op_geth.go [hunk 21]
```diff
@@ -226,7 +226,7 @@ func (d *OpGeth) CreatePayloadAttributes(txs ...*types.Transaction) (*eth.Payloa
 			if txBytes[i][0] == types.DepositTxType {
 				bin, err := derive.ToKromaDepositBytes(txBytes[i])
 				if err != nil {
-					return nil, fmt.Errorf("failed to convert DepositTx to KromaDepositTx: %w", err)
+					return nil, err
```
Comment (seolaoh) [inline op-e2e/op_geth.go:229]: Why did you change it?

Comment (Pangssu) [reply op-e2e/op_geth.go:229]: I requested it. `ToKromaDepositBytes` function already return this error.

**Review by seolaoh — APPROVED** [review]

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 22]
```diff
@@ -53,67 +57,175 @@ func TestMPTSourcesMatchSpec(t *testing.T) {
...
+	constructorArg := "0xA03c13C6597a0716D1525b7fDaD2fD95ECb49081"
```
Comment (sm-stack) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:190]: How do you think about changing these to mock, short bytecodes?

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 23]
```diff
@@ -53,67 +57,175 @@ func TestMPTSourcesMatchSpec(t *testing.T) {
...
+
+	// Validate the split between base bytecode and constructor argument
+	actualBytecode := actualDeploymentBytecodeHex[:len(bytecode)]
+	actualArg := actualDeploymentBytecodeHex[len(bytecode):]
+	require.Equal(t, bytecode, actualBytecode, "base bytecode invalid")
+	require.Equal(t, expectedEncodedArg, actualArg, "constructor argument invalid")
+}
+
+// Function to encode constructor arguments
+func encodeConstructorArg(arg string) (string, error) {
```
Comment (sm-stack) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:214]: How about moving this function to `kromampt_upgrade_transaction.go` and share it with this test file?

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 24]
```diff
@@ -30,14 +31,90 @@ var (
...
+// Function to create deployment bytecode with constructor arguments
+func createDeploymentBytecode(bytecode []byte, args ...interface{}) []byte {
+	// Define ABI arguments for the constructor
+	addressType, _ := abi.NewType("address", "", nil)
+	constructorArgs := abi.Arguments{
+		{
+			Type: addressType,
+		},
+	}
+	encodedArgs, _ := constructorArgs.Pack(args...)
```
Comment (sm-stack) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:108]: Don't we have to handle error here?

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 25]
```diff
@@ -30,14 +31,90 @@ var (
...
 )
 
-func KromaMPTNetworkUpgradeTransactions() ([]hexutil.Bytes, error) {
+// Bytecode mapping by ChainID
+type ChainConfig struct {
+	protocolVaultRecipient              common.Address
+	l1FeeVaultRecipient                 common.Address
+	SequencerFeeVaultDeploymentBytecode []byte
+	BaseFeeVaultDeploymentBytecode      []byte
+	L1FeeVaultDeploymentBytecode        []byte
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:42]: I think it would be nice to visualize the differences based on the constructor argument, so I'll keep that and minimize the other common parts.

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
...0000000083106107e0576d04ee2d6d415b85acef8100000000830492506020015b662386f26fc1000083106107fe57662386f26fc10000830492506010015b6305f5e1008310610816576305f5e100830492506008015b612710831061082a57612710830492506004015b6064831061083c576064830492506002015b600a8310610848576001015b92915050565b60005b83811015610869578181015183820152602001610851565b83811115610878576000848401525b50505050565b6000815180845261089681602086016020860161084e565b601f017fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe0169290920160200192915050565b6020815260006108db602083018461087e565b9392505050565b73ffffffffffffffffffffffffffffffffffffffff8416815263ffffffff8316602082015260606040820152600061091d606083018461087e565b95945050505050565b6000845161093881846020890161084e565b80830190507f2e000000000000000000000000000000000000000000000000000000000000008082528551610974816001850160208a0161084e565b6001920191820152835161098f81600284016020880161084e565b0160020195945050505050565b600082198211156109d6577f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b500190565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fdfea164736f6c634300080f000a")
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:43]: I wonder if we cannot utilize the bytecode from bindings

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+		case uint64:
+			argType, err = abi.NewType("uint64", "", nil)
+		case *big.Int:
+			argType, err = abi.NewType("uint256", "", nil)
+		case []byte:
+			argType, err = abi.NewType("bytes", "", nil)
+		case bool:
+			argType, err = abi.NewType("bool", "", nil)
+		default:
+			return nil, fmt.Errorf("unsupported argument type: %T", v)
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions.go:142]: Why not restrict only the expected types?
```suggestion
		default:
			return nil, fmt.Errorf("unexpected argument type: %T", v)
```

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+			name:    "Kroma Holesky configuration",
+			chainId: big.NewInt(params.KromaDevnetChainID),
+		},
+		{
+			name:    "Unsupported chain. Defaults to local devnet",
+			chainId: big.NewInt(9999),
+		},
+	}
+
+	for _, tt := range tests {
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:87]:
```suggestion
	for _, test := range tests {
```

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 29]
```diff
@@ -0,0 +1,175 @@
...
+			deploymentData, err := getDeploymentData(tt.chainId)
+			require.NoError(t, err)
+			require.Len(t, upgradeTxns, 8)
+
+			deployBaseFeeVaultSender, deployBaseFeeVault := toDepositTxn(t, upgradeTxns[1])
+			require.Equal(t, deployBaseFeeVaultSender, BaseFeeVaultDeployerAddress)
+			require.Equal(t, deployBaseFeeVaultSource.SourceHash(), deployBaseFeeVault.SourceHash())
+			require.Nil(t, deployBaseFeeVault.To())
+			require.Equal(t, uint64(1_000_000), deployBaseFeeVault.Gas())
+			require.Equal(t, fmt.Sprintf("0x%x", deploymentData.baseFeeVaultDeploymentBytecodeWithArg), hexutil.Bytes(deployBaseFeeVault.Data()).String())
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:100]: Same applys to below `fmt.Sprintf("0x%x", ~)`
```suggestion
			require.Equal(t, hexutil.Bytes(deploymentData.baseFeeVaultDeploymentBytecodeWithArg).String(), hexutil.Bytes(deployBaseFeeVault.Data()).String())
```

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 30]
```diff
@@ -0,0 +1,175 @@
...
+		{
+			name:    "Kroma Sepolia configuration",
+			chainId: big.NewInt(params.KromaSepoliaChainID),
+		},
+		{
+			name:    "Kroma Holesky configuration",
+			chainId: big.NewInt(params.KromaDevnetChainID),
+		},
+		{
+			name:    "Unsupported chain. Defaults to local devnet",
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:82]:
```suggestion
			name:    "LocalDevnet",
			chainId: big.NewInt(KromaLocalDevnetChainID),
		},
		{
			name:    "KromaMainnet",
			chainId: big.NewInt(params.KromaMainnetChainID),
		},
		{
			name:    "KromaSepolia",
			chainId: big.NewInt(params.KromaSepoliaChainID),
		},
		{
			name:    "KromaHolesky",
			chainId: big.NewInt(params.KromaDevnetChainID),
		},
		{
			name:    "UnsupportedChain",
```

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+			require.Equal(t, deploySequencerFeeVaultSender, SequencerFeeVaultDeployerAddress)
+			require.Equal(t, deploySequencerFeeVaultSource.SourceHash(), deploySequencerFeeVault.SourceHash())
+			require.Nil(t, deploySequencerFeeVault.To())
+			require.Equal(t, uint64(1_000_000), deploySequencerFeeVault.Gas())
+			require.Equal(t, fmt.Sprintf("0x%x", deploymentData.sequencerFeeVaultDeploymentBytecodeWithArg), hexutil.Bytes(deploySequencerFeeVault.Data()).String())
+		})
+	}
+
+	// test chainId-independent configurations
+	t.Run("ChainId-Independent Configurations", func(t *testing.T) {
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:119]:
```suggestion
	t.Run("ChainIDIndependent", func(t *testing.T) {
```

File: op-node/rollup/derive/kromampt_upgrade_transactions_test.go [hunk 32]
```diff
@@ -0,0 +1,175 @@
...
+		require.Equal(t, uint64(50_000), updateL1BlockProxy.Gas())
+		require.Equal(t, common.FromHex("0x3659cfe6000000000000000000000000cd7467a8926d13f8b41ea035ff3761326c822bad"), updateL1BlockProxy.Data())
+
+		updateBaseFeeVaultProxySender, updateBaseFeeVaultProxy := toDepositTxn(t, upgradeTxns[5])
+		require.Equal(t, updateBaseFeeVaultProxySender, common.Address{})
+		require.Equal(t, updateBaseFeeVaultProxySource.SourceHash(), updateBaseFeeVaultProxy.SourceHash())
+		require.NotNil(t, updateBaseFeeVaultProxy.To())
+		require.Equal(t, *updateBaseFeeVaultProxy.To(), predeploys.BaseFeeVaultAddr)
+		require.Equal(t, uint64(50_000), updateBaseFeeVaultProxy.Gas())
+		require.Equal(t, common.FromHex("3659cfe6000000000000000000000000db78bab44d9632e68348659dd47b4806ba276d89"), updateBaseFeeVaultProxy.Data())
```
Comment (seolaoh) [inline op-node/rollup/derive/kromampt_upgrade_transactions_test.go:146]:
```suggestion
		require.Equal(t, common.FromHex("3659cfe6000000000000000000000000db78bab44d9632e68348659dd47b4806ba276d89"), updateBaseFeeVaultProxy.Data())

		updateL1FeeVaultProxySender, updateL1FeeVaultProxy := toDepositTxn(t, upgradeTxns[6])
		require.Equal(t, updateL1FeeVaultProxySender, common.Address{})
		require.Equal(t, updateL1FeeVaultProxySource.SourceHash(), updateL1FeeVaultProxy.SourceHash())
		require.NotNil(t, updateL1FeeVaultProxy.To())
		require.Equal(t, *updateL1FeeVaultProxy.To(), predeploys.L1FeeVaultAddr)
		require.Equal(t, uint64(50_000), updateL1FeeVaultProxy.Gas())
		require.Equal(t, common.FromHex("3659cfe6000000000000000000000000d86e1a7c380f398bda3d598ee65c891ce5c3c8f0"), updateL1FeeVaultProxy.Data())

		updateSequencerFeeVaultProxySender, updateSequencerFeeVaultProxy := toDepositTxn(t, upgradeTxns[7])
		require.Equal(t, updateSequencerFeeVaultProxySender, common.Address{})
		require.Equal(t, updateSequencerFeeVaultProxySource.SourceHash(), updateSequencerFeeVaultProxy.SourceHash())
		require.NotNil(t, updateSequencerFeeVaultProxy.To())
		require.Equal(t, *updateSequencerFeeVaultProxy.To(), predeploys.SequencerFeeVaultAddr)
		require.Equal(t, uint64(50_000), updateSequencerFeeVaultProxy.Gas())
		require.Equal(t, common.FromHex("3659cfe600000000000000000000000013963c74d9c62d31ce3fcec0d46f6430a74ea79e"), updateSequencerFeeVaultProxy.Data())
```

**Review by sm-stack — APPROVED** [review]

File: op-node/rollup/derive/kromampt_upgrade_transactions.go (hunk 26 above)
Comment (sm-stack) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:43]: Won't it trigger problems for nodes who wanna derive the whole chain after MPT migration, if we change the bytecode of those contracts?

Comment (seolaoh) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:43]: Oh yeah I meant to bring the bytecode from current bindings, not from past deployments. I think this bytecode is from packages/contracts/deployments/kroma/ProtocolVault.json.

Comment (sm-stack) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:43]: Oh I see, nice catch!

Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go]: updated by bindings~!

**Review by seolaoh — APPROVED** [review]

File: op-node/rollup/derive/kromampt_upgrade_transactions.go [hunk 33]
```diff
@@ -0,0 +1,292 @@
...
...0000000000000000000000815260206004820152604a60248201527f4665655661756c743a207769746864726177616c20616d6f756e74206d75737460448201527f2062652067726561746572207468616e206d696e696d756d207769746864726160648201527f77616c20616d6f756e7400000000000000000000000000000000000000000000608482015260a40161026e565b6000479050806000808282546105a69190610707565b9091555050604080518281527f000000000000000000000000000000000000000000000000000000000000000073ffffffffffffffffffffffffffffffffffffffff166020820152338183015290517fc8a211cc64b6ed1b50595a9fcb1932b6d1e5a6e8ef15b60e5b1f988ea9086bba9181900360600190a1919050565b600080600080845160208601878a8af19695505050505050565b6000815180845260005b8181101561066457602081850181015186830182015201610648565b81811115610676576000602083870101525b50601f017fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe0169290920160200192915050565b6020815260006106bc602083018461063e565b9392505050565b73ffffffffffffffffffffffffffffffffffffffff8416815263ffffffff831660208201526060604082015260006106fe606083018461063e565b95945050505050565b60008219821115610741577f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b50019056fea164736f6c634300080f000a")
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:45]: Since a user who derives kroma through KromaMPT hardfork in the far future, not using snapshot, would have a trouble to successfully derive the chain when we use the bytecode in bindings which can be changed dynamically. So we hardcoded the bytecode in current bindings like a snapshot.

[the remark this answers is not in the document]
Comment (kangsorang) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:45]: 1. We've had this discussion before, and it took the form of embedding bytecode. https://github.com/kroma-network/kroma/pull/398#discussion_r1881229723
2. It would be nice to be able to verify via a hash or something, but does the hash value for the deployment byte code exist as an output here?

[the remark this answers is not in the document]
Comment (0xbenyun) [reply op-node/rollup/derive/kromampt_upgrade_transactions.go:45]: Thanks for explanation.  I just showed you the logic. Let's keep it as is

**Review by 0xbenyun — APPROVED** [review]
